File: uxsp/core/live.py
"""
UXSP Live Streaming Module (`uxsp.core.live`)

High-performance, zero-parsing symmetric encryption for live streaming (WebRTC video frames,
DataChannels, or raw WebSockets). 

This module bypasses the standard JSON `SecurePackage` envelope format to provide millisecond
latency suitable for 60 FPS video calls and real-time CCTV pipelines.
"""

from __future__ import annotations

import os
import struct

from uxsp.crypto.symmetric import KEY_SIZE, NONCE_SIZE, encrypt, decrypt
from uxsp.crypto.kdf import derive_key
from uxsp.core.replay import ReplayError


class LiveSession:
    """
    Manages a blazing-fast AES-GCM symmetric session key for real-time video/audio frames.
    """

    def __init__(self, key: bytes | None = None, session_id: bytes | None = None) -> None:
        """
        Initialize the session. If key is None, a new random 256-bit key is generated.
        """
        if key is None:
            self.key = os.urandom(KEY_SIZE)
        else:
            if len(key) != KEY_SIZE:
                raise ValueError(f"LiveSession key must be {KEY_SIZE} bytes.")
            self.key = bytes(key)
        self.session_id_bytes = session_id if isinstance(session_id, bytes) else (session_id.encode() if session_id else os.urandom(16))
        self._frame_count = 0
        self._last_seq = -1

    def _ratchet_key(self) -> None:
        self.key = derive_key(self.key, info=b"UXSP-live-ratchet", length=32)
        self._frame_count = 0

    @classmethod
    def generate(cls) -> LiveSession:
        """Create a new LiveSession with a fresh cryptographic key."""
        return cls()

    def encrypt_frame(self, frame: bytes | bytearray, metadata: bytes = b"") -> bytes:
        """
        Encrypt a raw binary frame (e.g. video codec output).
        Optionally attaches unencrypted (but mathematically authenticated) metadata.
        
        Format: [2-byte Metadata Length] [Metadata Bytes] [12-byte Nonce] [Ciphertext]
        """
        if len(metadata) > 65535:
            raise ValueError("Metadata too large (max 65535 bytes).")

        ad = self.session_id_bytes + metadata
        enc_dict = encrypt(bytes(frame), self.key, associated_data=ad)
        
        # Combine length + metadata + nonce + ciphertext
        meta_len = struct.pack(">H", len(metadata))
        result = meta_len + metadata + enc_dict["nonce"] + enc_dict["ciphertext"]
        
        self._frame_count += 1
        if self._frame_count >= 65536:
            self._ratchet_key()
            
        return result

    def decrypt_frame(self, encrypted_frame: bytes | bytearray, expected_seq: int | None = None) -> tuple[bytes, bytes]:
        """
        Decrypt a raw binary frame received over the wire.
        Returns a tuple: (decrypted_frame_bytes, unencrypted_metadata_bytes)
        """
        if len(encrypted_frame) < 2:
            raise ValueError("Encrypted frame is too small to contain length header.")
        
        meta_len = struct.unpack(">H", encrypted_frame[:2])[0]
        
        if len(encrypted_frame) < 2 + meta_len + NONCE_SIZE:
            raise ValueError("Encrypted frame is too small to contain metadata and nonce.")
        
        metadata = encrypted_frame[2 : 2 + meta_len]
        nonce = encrypted_frame[2 + meta_len : 2 + meta_len + NONCE_SIZE]
        ciphertext = encrypted_frame[2 + meta_len + NONCE_SIZE :]
        
        ad = self.session_id_bytes + metadata
        decrypted = decrypt(ciphertext, nonce, self.key, associated_data=ad)
        
        if expected_seq is not None:
            if expected_seq <= self._last_seq:
                raise ReplayError("Frame replay detected")
            self._last_seq = expected_seq
            
        self._frame_count += 1
        if self._frame_count >= 65536:
            self._ratchet_key()
            
        return decrypted, metadata

    @classmethod
    def extract_metadata(cls, encrypted_frame: bytes | bytearray) -> bytes:
        """
        Utility for routing servers (SFUs) to inspect the unencrypted metadata 
        attached to a frame without needing the decryption key.
        """
        if len(encrypted_frame) < 2:
            raise ValueError("Malformed frame: too short to contain metadata length.")
        meta_len = struct.unpack(">H", encrypted_frame[:2])[0]
        if len(encrypted_frame) < 2 + meta_len:
            raise ValueError("Malformed frame: too short to contain full metadata.")
        return encrypted_frame[2 : 2 + meta_len]

    def encrypt_voice_frame(
        self,
        frame: bytes | bytearray,
        *,
        codec: str = "opus",
        sample_rate: int = 48000,
        channels: int = 1,
        sequence: int = 0,
        is_muted: bool = False,
        metadata: bytes = b"",
    ) -> bytes:
        """
        Encrypt a live audio/voice frame with authenticated voice metadata.
        """
        import json

        audio_meta = {
            "type": "voice",
            "codec": codec,
            "sample_rate": sample_rate,
            "channels": channels,
            "sequence": sequence,
            "is_muted": is_muted,
        }
        if metadata:
            audio_meta["extra"] = metadata.hex()

        meta_bytes = json.dumps(audio_meta, separators=(",", ":")).encode("utf-8")
        return self.encrypt_frame(frame, metadata=meta_bytes)

    def decrypt_voice_frame(
        self,
        encrypted_frame: bytes | bytearray,
    ) -> tuple[bytes, dict[str, Any]]:
        """
        Decrypt a live voice audio frame and return (decrypted_audio_bytes, audio_metadata_dict).
        """
        import json

        decrypted_frame, meta_bytes = self.decrypt_frame(encrypted_frame)
        try:
            audio_meta = json.loads(meta_bytes.decode("utf-8"))
            if "extra" in audio_meta and isinstance(audio_meta["extra"], str):
                audio_meta["extra_bytes"] = bytes.fromhex(audio_meta["extra"])
        except Exception:
            audio_meta = {"raw_metadata": meta_bytes}
        return decrypted_frame, audio_meta


class LiveVoiceSession(LiveSession):
    """
    Manages an encrypted live voice/audio call session between peers.
    Supports zero-latency audio frame encryption (Opus, PCM, AAC)
    with authenticated audio call metadata (codec, sample rate, sequence, mute state).
    """

    def __init__(
        self,
        key: bytes | None = None,
        session_id: bytes | None = None,
        codec: str = "opus",
        sample_rate: int = 48000,
        channels: int = 1,
    ) -> None:
        super().__init__(key=key, session_id=session_id)
        self.codec = codec
        self.sample_rate = sample_rate
        self.channels = channels
        self.sequence = 0
        self.is_muted = False

    @classmethod
    def generate_voice(
        cls,
        codec: str = "opus",
        sample_rate: int = 48000,
        channels: int = 1,
    ) -> LiveVoiceSession:
        """Create a new LiveVoiceSession with fresh cryptographic key and call settings."""
        return cls(key=None, session_id=None, codec=codec, sample_rate=sample_rate, channels=channels)

    def mute(self) -> None:
        """Mute local audio transmission state."""
        self.is_muted = True

    def unmute(self) -> None:
        """Unmute local audio transmission state."""
        self.is_muted = False

    def next_sequence(self) -> int:
        """Increment and return the next audio packet sequence number."""
        self.sequence += 1
        return self.sequence

    def encrypt_voice_frame(  # type: ignore[override]
        self,
        frame: bytes | bytearray,
        *,
        codec: str | None = None,
        sample_rate: int | None = None,
        channels: int | None = None,
        sequence: int | None = None,
        is_muted: bool | None = None,
        metadata: bytes = b"",
    ) -> bytes:
        """
        Encrypt a live audio frame using current voice session parameters.
        """
        seq = sequence if sequence is not None else self.next_sequence()
        muted = is_muted if is_muted is not None else self.is_muted
        cd = codec or self.codec
        sr = sample_rate or self.sample_rate
        ch = channels or self.channels

        return super().encrypt_voice_frame(
            frame,
            codec=cd,
            sample_rate=sr,
            channels=ch,
            sequence=seq,
            is_muted=muted,
            metadata=metadata,
        )

File: sdks/js/src/live.ts
/**
 * UXSP Live Streaming Module (`live.ts`)
 * 
 * High-performance, zero-parsing symmetric encryption for WebRTC video frames,
 * DataChannels, or raw WebSockets in the browser.
 */

import { UXSPEnvelope, PublicCard } from "./types.js";
import { Identity } from "./identity.js";
import { seal, openSeal } from "./seal.js";
import { aesGcmEncrypt, aesGcmDecrypt } from "./crypto.js";

const KEY_SIZE = 32;
const NONCE_SIZE = 12;

export class LiveSession {
  public key: Uint8Array;

  constructor(key: Uint8Array) {
    if (key.byteLength !== KEY_SIZE) {
      throw new Error(`LiveSession key must be ${KEY_SIZE} bytes.`);
    }
    this.key = key;
  }

  /**
   * Encrypt a raw binary frame (e.g., from WebRTC Insertable Streams or DataChannel).
   * Optionally attaches unencrypted (but mathematically authenticated) metadata.
   * 
   * Format: [2-byte Metadata Length] [Metadata Bytes] [12-byte Nonce] [Ciphertext]
   */
  async encryptFrame(frame: Uint8Array, metadata?: Uint8Array): Promise<Uint8Array> {
    const meta = metadata || new Uint8Array(0);
    if (meta.byteLength > 65535) {
      throw new Error("Metadata too large (max 65535 bytes).");
    }

    const nonce = new Uint8Array(NONCE_SIZE);
    crypto.getRandomValues(nonce);

    const ciphertext = await aesGcmEncrypt(this.key, nonce, frame, meta);

    // Combine: [2-byte length] + [metadata] + [nonce] + [ciphertext]
    const payload = new Uint8Array(2 + meta.byteLength + nonce.byteLength + ciphertext.byteLength);
    
    // Write 2-byte length (Big Endian)
    const view = new DataView(payload.buffer, payload.byteOffset, payload.byteLength);
    view.setUint16(0, meta.byteLength, false);

    payload.set(meta, 2);
    payload.set(nonce, 2 + meta.byteLength);
    payload.set(ciphertext, 2 + meta.byteLength + nonce.byteLength);

    return payload;
  }

  /**
   * Decrypt a raw binary frame.
   * Returns both the decrypted frame and the unencrypted metadata.
   */
  async decryptFrame(encryptedFrame: Uint8Array): Promise<{ frame: Uint8Array; metadata: Uint8Array }> {
    if (encryptedFrame.byteLength < 2) {
      throw new Error("Encrypted frame is too small to contain length header.");
    }

    const view = new DataView(encryptedFrame.buffer, encryptedFrame.byteOffset, encryptedFrame.byteLength);
    const metaLen = view.getUint16(0, false);

    if (encryptedFrame.byteLength < 2 + metaLen + NONCE_SIZE) {
      throw new Error("Encrypted frame is too small to contain metadata and nonce.");
    }

    const metadata = encryptedFrame.slice(2, 2 + metaLen);
    const nonce = encryptedFrame.slice(2 + metaLen, 2 + metaLen + NONCE_SIZE);
    const ciphertext = encryptedFrame.slice(2 + metaLen + NONCE_SIZE);

    const frame = await aesGcmDecrypt(this.key, nonce, ciphertext, metadata);
    return { frame, metadata };
  }

  /**
   * Create a new LiveSession, returning the AES-GCM wrapper and the sealed envelope
   * to transmit to the peer via signaling.
   */
  static async create(
    sender: Identity,
    receiverCard: PublicCard
  ): Promise<{ envelope: UXSPEnvelope; session: LiveSession }> {
    const key = new Uint8Array(KEY_SIZE);
    crypto.getRandomValues(key);

    const envelope = await seal(sender, receiverCard, key);
    const session = new LiveSession(key);

    return { envelope, session };
  }

  /**
   * Accept an incoming LiveSession envelope from a peer, decrypting the 32-byte key.
   */
  static async accept(
    receiver: Identity,
    senderCard: PublicCard,
    envelope: UXSPEnvelope
  ): Promise<LiveSession> {
    const key = await openSeal(receiver, senderCard, envelope);
    return new LiveSession(key);
  }
}

export interface AudioMetadata {
  type: string;
  codec: string;
  sampleRate: number;
  channels: number;
  sequence: number;
  isMuted: boolean;
  extra?: string;
  extraBytes?: Uint8Array;
  [key: string]: any;
}

export class LiveVoiceSession extends LiveSession {
  public codec: string;
  public sampleRate: number;
  public channels: number;
  public sequence: number = 0;
  public isMuted: boolean = false;

  constructor(key: Uint8Array, codec = "opus", sampleRate = 48000, channels = 1) {
    super(key);
    this.codec = codec;
    this.sampleRate = sampleRate;
    this.channels = channels;
  }

  mute(): void {
    this.isMuted = true;
  }

  unmute(): void {
    this.isMuted = false;
  }

  nextSequence(): number {
    this.sequence += 1;
    return this.sequence;
  }

  async encryptVoiceFrame(
    frame: Uint8Array,
    options?: {
      codec?: string;
      sampleRate?: number;
      channels?: number;
      sequence?: number;
      isMuted?: boolean;
      metadata?: Uint8Array;
    }
  ): Promise<Uint8Array> {
    const seq = options?.sequence !== undefined ? options.sequence : this.nextSequence();
    const muted = options?.isMuted !== undefined ? options.isMuted : this.isMuted;
    const cd = options?.codec || this.codec;
    const sr = options?.sampleRate || this.sampleRate;
    const ch = options?.channels || this.channels;

    const audioMeta: Record<string, any> = {
      type: "voice",
      codec: cd,
      sample_rate: sr,
      channels: ch,
      sequence: seq,
      is_muted: muted,
    };

    if (options?.metadata && options.metadata.byteLength > 0) {
      audioMeta.extra = Array.from(options.metadata)
        .map((b) => b.toString(16).padStart(2, "0"))
        .join("");
    }

    const metaBytes = new TextEncoder().encode(JSON.stringify(audioMeta));
    return this.encryptFrame(frame, metaBytes);
  }

  async decryptVoiceFrame(
    encryptedFrame: Uint8Array
  ): Promise<{ frame: Uint8Array; audioMetadata: AudioMetadata }> {
    const { frame, metadata } = await this.decryptFrame(encryptedFrame);
    try {
      const metaText = new TextDecoder().decode(metadata);
      const parsed = JSON.parse(metaText);
      const audioMeta: AudioMetadata = {
        type: parsed.type || "voice",
        codec: parsed.codec || "opus",
        sampleRate: parsed.sample_rate || 48000,
        channels: parsed.channels || 1,
        sequence: parsed.sequence || 0,
        isMuted: Boolean(parsed.is_muted),
      };

      if (parsed.extra && typeof parsed.extra === "string") {
        const hex = parsed.extra;
        const match = hex.match(/.{1,2}/g);
        if (match) {
          audioMeta.extraBytes = new Uint8Array(match.map((byte: string) => parseInt(byte, 16)));
        }
      }

      return { frame, audioMetadata: audioMeta };
    } catch {
      return {
        frame,
        audioMetadata: {
          type: "raw",
          codec: "unknown",
          sampleRate: 0,
          channels: 0,
          sequence: 0,
          isMuted: false,
          extraBytes: metadata,
        },
      };
    }
  }

  static async createVoice(
    sender: Identity,
    receiverCard: PublicCard,
    options?: { codec?: string; sampleRate?: number; channels?: number }
  ): Promise<{ envelope: UXSPEnvelope; session: LiveVoiceSession }> {
    const key = new Uint8Array(KEY_SIZE);
    crypto.getRandomValues(key);

    const envelope = await seal(sender, receiverCard, key);
    const session = new LiveVoiceSession(
      key,
      options?.codec || "opus",
      options?.sampleRate || 48000,
      options?.channels || 1
    );

    return { envelope, session };
  }

  static async acceptVoice(
    receiver: Identity,
    senderCard: PublicCard,
    envelope: UXSPEnvelope,
    options?: { codec?: string; sampleRate?: number; channels?: number }
  ): Promise<LiveVoiceSession> {
    const key = await openSeal(receiver, senderCard, envelope);
    return new LiveVoiceSession(
      key,
      options?.codec || "opus",
      options?.sampleRate || 48000,
      options?.channels || 1
    );
  }
}

